nominal/sdk.py

```diff
@@ -0,0 +1,508 @@
+
+_AllowedFileExtensions = Literal[".csv", ".csv.gz", ".parquet"]
+
+__all__ = [
+    "NominalClient",
+    "Run",
+    "Dataset",
+    "Attachment",
+    "IntegralNanosecondsUTC",
+    "CustomTimestampFormat",
+    "NominalIngestError",
+    "NominalIngestFailed",
+]
+
+
+@dataclass(frozen=True)
+class Run:
+    rid: str
+    title: str
+    description: str
+    properties: Mapping[str, str]
+    labels: Sequence[str]
+    start: IntegralNanosecondsUTC
+    end: IntegralNanosecondsUTC | None
+    _auth_header: str = field(repr=False)
+    _run_client: scout.RunService = field(repr=False)
+
+    def add_datasets(self, datasets: Mapping[str, str]) -> None:
+        """Add datasets to this run.
+        Datasets map "ref names" (their name within the run) to dataset RIDs.
+            The same type of datasets should use the same ref name across runs,
+            since checklists and templates use ref names to reference datasets.
+            The RIDs are retrieved from creating or getting a `Dataset` object.
+        """
+        data_sources = {
+            ref_name: scout_run_api.CreateRunDataSource(
+                data_source=scout_run_api.DataSource(dataset=rid),
+                series_tags={},
+                offset=None,  # TODO(alkasm): support per-dataset offsets
+            )
+            for ref_name, rid in datasets.items()
+        }
+        self._run_client.add_data_sources_to_run(self._auth_header, data_sources, self.rid)
+
+    def list_datasets(self) -> Iterable[tuple[str, str]]:
+        """List the datasets associated with this run.
+        Yields (ref_name, dataset_rid) pairs.
+        """
+        run = self._run_client.get_run(self._auth_header, self.rid)
+        for ref_name, source in run.data_sources.items():
+            if source.data_source.type == "dataset":
+                dataset_rid = cast(str, source.data_source.dataset)
+                yield (ref_name, dataset_rid)
+
+    def add_attachments(self, attachment_rids: Iterable[str]) -> None:
+        """Add attachments that have already been uploaded to this run."""
+        request = scout_run_api.UpdateAttachmentsRequest(
+            attachments_to_add=list(attachment_rids), attachments_to_remove=[]
+        )
+        self._run_client.update_run_attachment(self._auth_header, request, self.rid)
+
+    def remove_attachments(self, attachment_rids: Iterable[str]) -> None:
+        """Remove attachments from this run.
+        Does not remove the attachments from Nominal.
+        """
+        request = scout_run_api.UpdateAttachmentsRequest(
+            attachments_to_add=[], attachments_to_remove=list(attachment_rids)
+        )
+        self._run_client.update_run_attachment(self._auth_header, request, self.rid)
+
+    def replace(
+        self,
+        *,
+        title: str | None = None,
+        description: str | None = None,
+        properties: Mapping[str, str] | None = None,
+        labels: Sequence[str] | None = None,
+    ) -> Self:
```

Comment:
Yeah, I waffled on this myself. Since you're also bringing it up I think I'll mutate the instance. 

Initially I wrote all the attributes with a @property decorator so I could make them "immutable" to the user but still update them. But then I went to frozen dataclasses for simplicity. I can still mutate the members in the object using __setattr__ even if using a frozen dataclass.

Reply:
should I rename them back to `update()` then? maybe more intuitive?